### PR TITLE
Leapfrog

### PR DESCRIPTION
**Pull request by eggrobin**
looks like we'll be needing a bibliography soon...

### CODE REVIEW

File: integrators/symplectic_partitioned_runge_kutta_integrator.hpp [hunk 1]
```diff
@@ -11,11 +12,26 @@ using base::not_null;
...
+  kFirstBVanishes,
+  kLastAVanishes,
+};
+
 template<typename Position, typename Momentum>
 class SPRKIntegrator : public SymplecticIntegrator<Position, Momentum> {
  public:
-  using Coefficients = typename SymplecticIntegrator<Position,
-                                                     Momentum>::Coefficients;
+  using Scheme = typename SymplecticIntegrator<Position, Momentum>::Scheme;
```
Comment (pleroy) [inline integrators/symplectic_partitioned_runge_kutta_integrator.hpp]: Alphabetize the type names.

File: integrators/symplectic_partitioned_runge_kutta_integrator.hpp [hunk 2]
```diff
@@ -24,9 +40,79 @@ class SPRKIntegrator : public SymplecticIntegrator<Position, Momentum> {
   SPRKIntegrator();
   ~SPRKIntegrator() override = default;
 
-  std::vector<std::vector<double>> const& Order5Optimal() const;
+  // First-same-as-last (FSAL) methods with k stages require k - 1 force
+  // and velocity evaluations for sparse output.  For dense output, methods
+  // with synchronous momenta require k - 1 force evalutations and k velocity
```
Comment (pleroy) [inline integrators/symplectic_partitioned_runge_kutta_integrator.hpp]: s/t//

File: integrators/symplectic_partitioned_runge_kutta_integrator.hpp [hunk 3]
```diff
@@ -24,9 +40,79 @@ class SPRKIntegrator : public SymplecticIntegrator<Position, Momentum> {
...
+  // evaluations, methods with synchronous positions require k force evaluations
+  // evaluations and k - 1 velocity evaluations.
```
Comment (pleroy) [inline integrators/symplectic_partitioned_runge_kutta_integrator.hpp]: Stuttering.

File: integrators/symplectic_partitioned_runge_kutta_integrator.hpp [hunk 4]
```diff
@@ -36,13 +122,46 @@ class SPRKIntegrator : public SymplecticIntegrator<Position, Momentum> {
...
+  };
+
+  template<VanishingCoefficients vanishing_coefficients_,
+           typename AutonomousRightHandSideComputation,
+           typename RightHandSideComputation>
+  void SolveOptimized(RightHandSideComputation compute_force,
+                      AutonomousRightHandSideComputation compute_velocity,
+                      Parameters const& parameters,
+                      not_null<std::vector<SystemState>*> const solution) const;
+
```
Comment (pleroy) [inline integrators/symplectic_partitioned_runge_kutta_integrator.hpp]: Remove empty line.

File: integrators/symplectic_partitioned_runge_kutta_integrator.hpp [hunk 5]
```diff
@@ -36,13 +122,46 @@ class SPRKIntegrator : public SymplecticIntegrator<Position, Momentum> {
...
+  // a coefficients.
+  // If |vanishing_coefficients_ == kLastAVanishes|, we do not store the last a
+  // coefficient, and the first entry of |b_| is the sum of the first and last b
+  // coefficients.
   std::vector<double> a_;
   std::vector<double> b_;
 
-  // The weights.
+  // The position weights.  Note that the first c coefficient is not stored if
+  // |vanishing_coefficients_ == kFirstBVanishes|.
```
Comment (pleroy) [inline integrators/symplectic_partitioned_runge_kutta_integrator.hpp]: Is there something analog (eg last not stored) if kLastAVanishes?

File: integrators/symplectic_partitioned_runge_kutta_integrator_body.hpp [hunk 6]
```diff
@@ -17,12 +45,106 @@ using quantities::Quotient;
...
+typename SPRKIntegrator<Position, Momentum>::Scheme const&
+SPRKIntegrator<Position, Momentum>::McLachlanAtela1992Order2Optimal() const {
+  static Scheme const scheme = {
+      { 0.7071067811865475244,
+        0.2928932188134524756},
+      { 0.2928932188134524756,
+        0.7071067811865475244}};
+  return scheme;
+}
+
```
Comment (pleroy) [inline integrators/symplectic_partitioned_runge_kutta_integrator_body.hpp]: Remove empty line.

File: integrators/symplectic_partitioned_runge_kutta_integrator_body.hpp [hunk 7]
```diff
@@ -35,23 +157,319 @@ SPRKIntegrator<Position, Momentum>::Order5Optimal() const {
...
+                             coefficients[1].end() - 1);
+    b_.front() += first_same_as_last_->last;
+    stages_ = b_.size();
+    CHECK_EQ(stages_, a_.size());
+  } else {
+    vanishing_coefficients_ = kNone;
+    a_ = coefficients[0];
+    b_ = coefficients[1];
+    stages_ = b_.size();
+    CHECK_EQ(stages_, a_.size());
```
Comment (pleroy) [inline integrators/symplectic_partitioned_runge_kutta_integrator_body.hpp]: Clear first_same_as_last.

File: integrators/symplectic_partitioned_runge_kutta_integrator_body.hpp [hunk 8]
```diff
@@ -35,23 +157,319 @@ SPRKIntegrator<Position, Momentum>::Order5Optimal() const {
...
+  }
 
   // Runge-Kutta time weights.
   c_.resize(stages_);
-  c_[0] = 0.0;
+  if (vanishing_coefficients_ == kFirstBVanishes) {
+    c_[0] = first_same_as_last_->first;
+  } else {
+    c_[0] = 0;
```
Comment (pleroy) [inline integrators/symplectic_partitioned_runge_kutta_integrator_body.hpp]: 0.0

File: integrators/symplectic_partitioned_runge_kutta_integrator_body.hpp [hunk 9]
```diff
@@ -63,6 +481,33 @@ void SPRKIntegrator<Position, Momentum>::Solve(
...
+      LOG(FATAL) << "Invalid vanishing coefficients";
+  }
+}
+
+template<typename Position, typename Momentum>
+template<VanishingCoefficients vanishing_coefficients,
+         typename AutonomousRightHandSideComputation,
+         typename RightHandSideComputation>
+void SPRKIntegrator<Position, Momentum>::SolveOptimized(
+      RightHandSideComputation compute_force,
```
Comment (pleroy) [inline integrators/symplectic_partitioned_runge_kutta_integrator_body.hpp]: const (2x)

File: integrators/symplectic_partitioned_runge_kutta_integrator_body.hpp [hunk 10]
```diff
@@ -143,28 +586,57 @@ void SPRKIntegrator<Position, Momentum>::Solve(
       (*Δpstage_current)[k] = Momentum();
       q_stage[k] = q_last[k].value;
     }
+
```
Comment (pleroy) [inline integrators/symplectic_partitioned_runge_kutta_integrator_body.hpp]: Remove blank line.

File: integrators/symplectic_partitioned_runge_kutta_integrator.hpp [hunk 11]
```diff
@@ -36,13 +122,46 @@ class SPRKIntegrator : public SymplecticIntegrator<Position, Momentum> {
              not_null<std::vector<SystemState>*> const solution) const;
 
  private:
+  struct FirstSameAsLast {
+    double first;
+    double last;
+  };
+
+  template<VanishingCoefficients vanishing_coefficients_,
```
Comment (pleroy) [inline integrators/symplectic_partitioned_runge_kutta_integrator.hpp]: No _ and fix all the references.

File: integrators/symplectic_partitioned_runge_kutta_integrator_test.cpp [hunk 12]
```diff
@@ -32,36 +31,181 @@ using quantities::SIUnit;
...
+// This allows the test output to be legible, i.e.,
+// "where GetParam() = Leapfrog" rather than
+// "where GetParam() = n-byte object <hex>"
+std::ostream& operator<<(std::ostream& stream, SPRKTestableProperties param) {
+  return stream << param.name;
+}
+
+std::vector<SPRKTestableProperties> Instances() {
+  return {
+      {&Integrator::Leapfrog, "Leapfrog", 2, 0.4 * Second,
```
Comment (pleroy) [inline integrators/symplectic_partitioned_runge_kutta_integrator_test.cpp]: Use a macro to stringize the 2nd parameter.

File: integrators/symplectic_partitioned_runge_kutta_integrator_test.cpp [hunk 13]
```diff
@@ -175,47 +325,67 @@ TEST_F(SPRKTest, Convergence) {
...
+  log_step_sizes.reserve(step_sizes);
+  std::vector<double> log_p_errors;
+  log_step_sizes.reserve(step_sizes);
   for (int i = 0; i < step_sizes; ++i, parameters_.Δt /= step_reduction) {
     integrator_.Solve(&ComputeHarmonicOscillatorForce,
                       &ComputeHarmonicOscillatorVelocity,
                       parameters_, &solution_);
-    log_step_sizes[i] = std::log10(parameters_.Δt / SIUnit<Time>());
-    log_q_errors[i] = std::log10(
+    double log_q_error = std::log10(
```
Comment (pleroy) [inline integrators/symplectic_partitioned_runge_kutta_integrator_test.cpp]: const

File: integrators/symplectic_partitioned_runge_kutta_integrator_test.cpp [hunk 14]
```diff
@@ -175,47 +325,67 @@ TEST_F(SPRKTest, Convergence) {
...
         std::abs(solution_[0].positions[0].value / SIUnit<Length>() -
                  Cos(solution_[0].time.value *
                      SIUnit<AngularFrequency>())));
-    log_p_errors[i] = std::log10(
+    double log_p_error = std::log10(
```
Comment (pleroy) [inline integrators/symplectic_partitioned_runge_kutta_integrator_test.cpp]: const

File: integrators/symplectic_partitioned_runge_kutta_integrator_test.cpp [hunk 15]
```diff
@@ -175,47 +325,67 @@ TEST_F(SPRKTest, Convergence) {
...
+    log_q_errors.push_back(log_q_error);
+    log_p_errors.push_back(log_p_error);
   }
   double const q_convergence_order = Slope(log_step_sizes, log_q_errors);
   double const q_correlation =
       PearsonProductMomentCorrelationCoefficient(log_step_sizes, log_q_errors);
+  LOG(INFO) << GetParam();
   LOG(INFO) << "Convergence order in q : " << q_convergence_order;
   LOG(INFO) << "Correlation            : " << q_correlation;
+#if 1
```
Comment (pleroy) [inline integrators/symplectic_partitioned_runge_kutta_integrator_test.cpp]: s/1/0/

File: integrators/symplectic_partitioned_runge_kutta_integrator.hpp [hunk 16]
```diff
@@ -24,9 +40,79 @@ class SPRKIntegrator : public SymplecticIntegrator<Position, Momentum> {
...
+  Scheme const& McLachlanAtela1992Order2Optimal() const;
+  // Third order, 3 stages.
+  // Coefficients from Ruth (1983), A canonical integration technique,
+  // https://accelconf.web.cern.ch/accelconf/p83/PDF/PAC1983_2669.PDF.
+  Scheme const& Ruth1983() const;
+  // Third order, 3 stages.  This method minimizes the error constant.
+  // Coefficients from McLachlan and Atela (1992),
+  // The accuracy of symplectic integrators, equation 3.3.
+  // http://eaton.math.rpi.edu/CSUMS/Papers/Symplectic/McLachlan_Atela_92.pdf.
+  // NOTE(egg): the coefficients give in table 2 for this integrator are
```
Comment (eggrobin) [inline integrators/symplectic_partitioned_runge_kutta_integrator.hpp]: given

File: integrators/symplectic_partitioned_runge_kutta_integrator_test.cpp (hunk 15 above)
Comment (eggrobin) [inline integrators/symplectic_partitioned_runge_kutta_integrator_test.cpp]: Done.

File: integrators/symplectic_partitioned_runge_kutta_integrator_test.cpp (hunk 14 above)
Comment (eggrobin) [inline integrators/symplectic_partitioned_runge_kutta_integrator_test.cpp]: Done.

File: integrators/symplectic_partitioned_runge_kutta_integrator_test.cpp (hunk 13 above)
Comment (eggrobin) [inline integrators/symplectic_partitioned_runge_kutta_integrator_test.cpp]: Done.

File: integrators/symplectic_partitioned_runge_kutta_integrator_test.cpp (hunk 12 above)
Comment (eggrobin) [inline integrators/symplectic_partitioned_runge_kutta_integrator_test.cpp]: Done.

File: integrators/symplectic_partitioned_runge_kutta_integrator_body.hpp (hunk 10 above)
Comment (eggrobin) [inline integrators/symplectic_partitioned_runge_kutta_integrator_body.hpp]: Done.

File: integrators/symplectic_partitioned_runge_kutta_integrator_body.hpp (hunk 9 above)
Comment (eggrobin) [inline integrators/symplectic_partitioned_runge_kutta_integrator_body.hpp]: Done.

File: integrators/symplectic_partitioned_runge_kutta_integrator_body.hpp (hunk 8 above)
Comment (eggrobin) [inline integrators/symplectic_partitioned_runge_kutta_integrator_body.hpp]: Done.

File: integrators/symplectic_partitioned_runge_kutta_integrator_body.hpp (hunk 7 above)
Comment (eggrobin) [inline integrators/symplectic_partitioned_runge_kutta_integrator_body.hpp]: Done.

File: integrators/symplectic_partitioned_runge_kutta_integrator_body.hpp (hunk 6 above)
Comment (eggrobin) [inline integrators/symplectic_partitioned_runge_kutta_integrator_body.hpp]: Done.

File: integrators/symplectic_partitioned_runge_kutta_integrator.hpp (hunk 5 above)
Comment (eggrobin) [inline integrators/symplectic_partitioned_runge_kutta_integrator.hpp]: No. I added a huge blurb of comments at the top, this should become clear.

File: integrators/symplectic_partitioned_runge_kutta_integrator.hpp (hunk 4 above)
Comment (eggrobin) [inline integrators/symplectic_partitioned_runge_kutta_integrator.hpp]: Done.

File: integrators/symplectic_partitioned_runge_kutta_integrator.hpp (hunk 11 above)
Comment (eggrobin) [inline integrators/symplectic_partitioned_runge_kutta_integrator.hpp]: Done.

File: integrators/symplectic_partitioned_runge_kutta_integrator.hpp (hunk 16 above)
Comment (eggrobin) [inline integrators/symplectic_partitioned_runge_kutta_integrator.hpp]: Done.

File: integrators/symplectic_partitioned_runge_kutta_integrator.hpp (hunk 3 above)
Comment (eggrobin) [inline integrators/symplectic_partitioned_runge_kutta_integrator.hpp]: Done.

File: integrators/symplectic_partitioned_runge_kutta_integrator.hpp (hunk 2 above)
Comment (eggrobin) [inline integrators/symplectic_partitioned_runge_kutta_integrator.hpp]: Done.

File: integrators/symplectic_partitioned_runge_kutta_integrator.hpp (hunk 1 above)
Comment (eggrobin) [inline integrators/symplectic_partitioned_runge_kutta_integrator.hpp]: Done.

**Comment by eggrobin** [comment]
`const` functors do not work, because the `operator()` of the functor need not be `const`.
Indeed in our case, we sometimes pass [the result of `std::bind`](http://en.cppreference.com/w/cpp/utility/functional/bind#Return_value), whose `operator()` is not const (that makes sense since `std::bind` may wrap a functor whose `operator()` is itself not `const`).

**Comment by eggrobin** [comment]
Actually it may be a good idea to pass functors by `&&` to avoid copies, but that's a job for another CL.

**Pull request merged**